### ISSUE TITLE
Multiple copies of Linked Abilities incorrectly see each other

### ISSUE THREAD

**Issue by ssk97** [magefree/mage#12752]
Right now, Linked Abilities are implemented by storing information on the card/permanent in a manner accessible to all other abilities. This mostly works, but has problems when linked abilities are granted to other cards, which can cause separate abilities to look up their information using the same string. (This information is stored in many different locations: costs tags, exile zone names, `game.getState().setValue`, etc.)

For Costs Tags, one example of this is [[Zinnia, Valley's Voice]], where if casting a spell that already has Offspring should make separate triggers and offspring for each cost paid. This also applies to Kicker and Squad (though those currently have no cards where that matters except for the un-set [[Wicker Picker]]) and potentially others.

I think we need some sort of more general way to attach linked abilities to each other, not sure exactly how. This linkage must remain across deep copy effects, but must also create a new link when copying the card it's on (such as by making a non-legendary copy of Zinnia). I don't currently have any solution here, I just wanted to bring the issue up and discuss. For more information on the correct rules, see [rule 607 Linked Abilities](https://yawgatog.com/resources/magic-rules/#R607).

**Comment by JayDi85** [comment]
Game state and watchers with stored data already has all tools. No needs to link java objects -- it's not compatible with game engine (object copied all around, abilities get's new ids on trigger, activate, etc).

Card id + zcc + searching key are good. There are also originalId logic (if you need really link two abilities and support multiple instances), but it's a rare usage and I don't recommend to use it without real needs.

**Comment by ssk97** [comment]
> Card id + zcc + searching key are good. There are also originalId logic (if you need really link two abilities and support multiple instances), but it's a rare usage and I don't recommend to use it without real needs.

"it's a rare usage" is true in that it is rare for the game state to be such that it matters, but there are several abilities where it can unexpectedly show up and the current behavior is wrong. As an example, have [[Izzet Chemister]] and [[Experiment Kraj]]. If you flicker the Chemister, Kraj should lose track of the previously cards it exiled.

Another example would be a card with "Choose a color", then becomes a copy of a card that also has a "choose a color" ability. Current XMage uses the same color for the 2nd, but it should not.

Implementing originalId logic everywhere might be able to handle those issues, but it's quite complex to handle correctly, and must be custom re-implemented every time it's used.

**Comment by JayDi85** [comment]
> Izzet see exiled cards

Yep, that’s wrong usage of xmage’s exile zones (it was a standard code style for years — use shared source related exile zone instead ability related). Card must use some unique prefix for a zone name, so only linked ability will access to it. Izzet don’t have ruling for ref but [[Admonition Angel]] has:

> Admonition Angel's landfall ability and its last ability are linked. The last ability refers only to cards exiled with its landfall ability.
(2010-03-01)

BTW same problem for other effects with standard names for a keys like “damagedPermanent”, etc. It can be accessed by wrong effects in theory. The only reason it’s a not popular bug — there is a little amount of cards and mechanics with massive abilities combine (hello mutate).

**Comment by ssk97** [comment]
> Yep, that’s wrong usage of xmage’s exile zones (it was a standard code style for years — use shared source related exile zone instead ability related). Card must use some unique prefix for a zone name, so only linked ability will access to it.

Even if you add a unique prefix, the above interaction still wouldn't work since both the original and the post-flicker ability would use the same prefix and thus see the same exile zone. This is wrong (assuming I understood the rules correctly. I haven't confirmed with a judge that the new Chemister permanent gives the same Kraj new abilities).

**Comment by JayDi85** [comment]
> Even if you add a unique prefix, the above interaction still wouldn't work since both the original and the post-flicker ability would use the same prefix and thus see the same exile zone

It must work with flicker too. Zone name = ability prefix + card/permanent id + zcc. There are special methods to find source zone with zcc offset (e.g. to support “on battlefield leave” logic).

**Comment by ssk97** [comment]
> > Even if you add a unique prefix, the above interaction still wouldn't work since both the original and the post-flicker ability would use the same prefix and thus see the same exile zone
> 
> It must work with flicker too. Zone name = ability prefix + card/permanent id + zcc. There are special methods to find source zone with zcc offset (e.g. to support “on battlefield leave” logic).

Ability prefix: same

Permanent ID+ZCC: same because the Experiment Kraj is the card with the ability and it has stayed on the battlefield the whole time, its zcc has not changed. The Chemister the ability is being copied from has a new zcc, but that's not helpful here.

**Comment by JayDi85** [comment]
I don't understand you.

Izzet + Kraj use case:
1. Exile cards by Izzet;
2. Put counter and give ability to Kraj;
3. Exile cards by Kraj and gained Izzet's ability;
4. Exile cards by Kraj and gained another's ability (for example);

Results on good implementation with existing tools (prefix + source zone + zcc):
  * Izzet view:
    * own exiled cards -- will see due same source zone, ability prefix and zcc;
    * Kraj's exiled cards by Izzet's gained ability  -- will NOT see due diff source zone;
    * Kraj's exiled cards by another's gained ability  -- will NOT see due diff source zone;
    * after blink - will NOT see any exiled cards due diff zcc;
  * Kraj's view:
    * own exiled cards by Izzet's gained ability -- will see due same source zone, ability prefix and zcc;
    * own exiled cards by other's gained ability -- will NOT see due diff ability prefix;
    * Izzet's exiled cards -- will NOT see due diff source zone;
    * after blink - will NOT see any exiled cards due diff zcc;

**Comment by ssk97** [comment]
Ah, my apologies, I wasn't clear enough in the scenario. Start with Izzet Chemister and Experiment Kraj on the battlefield.

1) Put a counter on Izzet, Kraj gains Izzet's Abilities
2) Kraj activates Izzet's 1st ability and exiles a card
3) Blink Izzet, Kraj loses abilities
4) Put a counter on Izzet, Kraj gains the new Izzet's Abilities
5) Kraj activates Izzet's 2nd ability

In step 5 there should be no cards found to be used, since it should be a new ability, copied from a new permanent. Izzet has a new zcc, but Kraj does not. AFAIK there is no reasonable way to have this interaction be correct with the current tools.

**Comment by JayDi85** [comment]
Well, I see. It's a weird use case to "blink" a single ability instead whole object. It's incompatible with current "copy abilities from existing objects" logic (current copy logic keep originalId to support linked abilities). E.g. require some type of ability's zcc.

BUT I think that's whole thing with ability's zcc is wrong -- it's SAME ability by MTG rules cause it ref to printed:
> 607.1a An ability printed on an object within another ability that grants that ability to that object is considered to be “printed on” that object for these purposes.

If not then any [[Blood Moon]] or other "remove all abilities" must broke all choices and links after reset abilities to normal state.

**Comment by JayDi85** [comment]
Yes, originalId as prefix must work with two independent pairs if linked abilities (exile id = prefix/original + object id + zcc)

**Comment by ssk97** [comment]
> Yes, originalId as prefix must work with two independent pairs if linked abilities (exile id = prefix/original + object id + zcc)

Currently, this breaks if the card is copied - that causes all originalIds to be reset to new ones, and there's no way to update the linked ability's stored originalId when they're changed during that copy.

**Comment by sneddigrolyat** [comment]
Hi all — I opened PR #14625 to address this issue to help implement Zinnia, Valley's Voice. 

This PR generalizes linked-ability tracking so costs paid on the stack stay correctly associated with the resulting ETB/linked abilities, including in copied spell scenarios. The key change is introducing a stable linkage id on abilities (shared across linked abilities and preserved through copying/re-application), and using that linkage id to build cost-tag keys. This prevents the “activation tag lost” cases that previously broke Offspring and other linked-ability mechanics when abilities were regenerated or spell copies were created.

How it works:
- `AbilityImpl` now carries a `linkageId` and exposes `getLinkageId()`. This id is propagated to sub-abilities.
- New helper `newIdKeepingLinkage()` lets abilities get fresh ids while preserving linkage across apply-effects cycles, so linked abilities remain associated even after ability re-creation.
- Cost tags for Offspring are now keyed using a linked-cost tag (via `CardUtil.getLinkedCostTag`), which is based on the linkage id instead of the old `originalId`. That removes the mismatch when abilities were copied or regenerated.
- On spell resolution, any “other abilities” granted on the stack are captured in `GameState.enteringAbilities` keyed to the next zone change counter (MageObjectReference). These are re-applied during `GameState.applyEnteringAbilities()` after the battlefield reset, so granted Offspring/Squad-style abilities persist through ETB even if the permanent’s printed abilities get stripped or the spell was copied.
- The previous Offspring delayed-trigger workaround is removed; Offspring now relies on the linkage-based cost tags being present at ETB. This reduces special-case logic and keeps Offspring aligned with other linked-ability mechanics.

Live gameplay tests (all passed):
1. Noncreature spell with Zinnia, Valley’s Voice: cast Lightning Bolt or Opt, no Offspring prompt.
2. Silvercoat Lion: exactly one Offspring prompt from Zinnia, Valley’s Voice.
3. Prosperous Bandit: two prompts (native Offspring + Zinnia, Valley’s Voice), outcomes match choices.
4. Countered creature: pay Offspring then counter with Counterspell/Cancel, no token created.
5. Elvish Visionary with Offspring: two ETB draws.
6. Combat/stack: Giant Growth + Lightning Bolt with Doom Blade responses, priority and resolution correct.
7. Zone transitions: Unsummon / Path to Exile / Doom Blade on original vs token behave correctly.
8. Disconnect/reconnect: state consistent.
9. Fast match flow: Colossal Dreadmaw + Shivan Dragon finish, next game/match state correct.
10. Humility in response: cast Iridescent Vinelasher, respond with Humility (via Vedalken Orrery), no Offspring token created.
11. Squad copy (Double Major): Arco-Flagellant paid Squad once, copy preserves Squad payment (4 total).
12. Squad ETB trigger copy (Strionic Resonator): Arco-Flagellant paid Squad once, copied trigger uses paid count (3 total).
13. Offspring spell copy (Double Major): Prosperous Bandit paid Offspring once, copy preserves Offspring payment (4 total).
14. Token copy of Offspring permanent: Cackling Counterpart on Prosperous Bandit creates only one token, no self-copy Offspring.
15. Zinnia removed before ETB: Path to Exile on Zinnia, Valley’s Voice in response, no Offspring token.

I’m new to the codebase and would appreciate more experienced review of this linkage approach — whether the generalized linkage id/cost-tag strategy is sound and if there are edge cases or unintended interactions I’ve missed. I’m open to any feedback or alternative directions.

**Comment by ssk97** [comment]
_edit: I should've posted this in the PR itself, if you respond do so in there_

Here's my first glance at things:

In general, storing references to abilities directly is bad because deep copies will make new, disconnected copies instead of keeping the actual abilities connected. This also creates problems where the unit tests don't fully match actual gameplay, since the unit tests don't do rollbacks whereas gameplay often does (every time someone taps mana wrong and does an undo).

Why are you checking for `AbilityImpl` instead of just adding the functions to the Ability interface?

You have a bunch of code in `PermanentCard`, does it work correctly with tokens (copied spells)? Does your Zinnia work correctly if you make a non-legendary copy of it on the battlefield?
